safehttp/plugins/xsrf/html/xsrf.go

```diff
@@ -0,0 +1,160 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xsrfhtml
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/google/go-safeweb/safehttp/plugins/xsrf"
+
+	"github.com/google/go-safeweb/safehttp"
+	"golang.org/x/net/xsrftoken"
+)
+
+const (
+	// TokenKey is the form key used when sending the token as part of POST
+	// request.
+	TokenKey    = "xsrf-token"
+	cookieIDKey = "xsrf-cookie"
+)
+
+// Interceptor implements XSRF protection.
+type Interceptor struct {
+	// SecretAppKey uniquely identifies each registered service and should have
+	// high entropy as it is used for generating the XSRF token.
+	SecretAppKey string
+}
+
+var _ safehttp.Interceptor = &Interceptor{}
+
+type tokenCtxKey struct{}
+
+// Token extracts the XSRF token from the incoming request. If it is not
+// present, it returns a non-nil error.
+func Token(r *safehttp.IncomingRequest) (string, error) {
+	tok := r.Context().Value(tokenCtxKey{})
+	if tok == nil {
+		return "", errors.New("xsrf token not found")
+	}
+	return tok.(string), nil
+}
+
+func addCookieID(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
+	buf := make([]byte, 20)
+	if _, err := rand.Read(buf); err != nil {
+		return nil, fmt.Errorf("crypto/rand.Read: %v", err)
+	}
+
+	c := safehttp.NewCookie(cookieIDKey, base64.StdEncoding.EncodeToString(buf))
+	c.SetSameSite(safehttp.SameSiteStrictMode)
+	if err := w.SetCookie(c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Before should be executed before directing the safehttp.IncomingRequest to
+// the handler to ensure it is not part of a Cross-Site Request
+// Forgery attack.
+//
+// On first user visit through a state preserving request (GET, HEAD or
+// OPTIONS), a nonce-based cookie will be set in the response as a way to
+// distinguish between users and prevent pre-login XSRF attacks. The cookie will
+// be used in the token generation and verification algorithm and is expected to
+// be present in all subsequent incoming requests.
+//
+// For every authorized request, the interceptor will also generate a
+// cryptographically-safe XSRF token using the appKey, the cookie and the path
+// visited. This can be later extracted using Token and should be injected as a
+// hidden input field in HTML forms.
+//
+// In case of state changing requests (all except GET, HEAD and OPTIONS), the
+// interceptor checks for the presence of the XSRF token in the request body
+// (expected to have been injected) and validates it.
+func (it *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, _ safehttp.InterceptorConfig) safehttp.Result {
+	needsValidation := !xsrf.StatePreserving(r)
```

Comment:
Please make this behave as the other code: as a first thing we do what we'll need to do every time: generate the token and add it to the context.

Then if the request is state preserving we just return notwritten, otherwise we continue with the checks.

The code below has a lot of nested ifs and this makes it hard to follow.

Reply:
Since this PR focuses on the Angular XSRF plugin, I will do it in a follow-up PR. Otherwise, I think it will make this PR harder to review since it will have some unrelated changes. Is that okay?

Reply:
Sounds good to me.